### PR TITLE
[mlir][Transforms][NFC] Move `ReconcileUnrealizedCasts` implementation

### PR DESCRIPTION
**Pull request by matthias-springer**
Move the implementation of `ReconcileUnrealizedCasts` to `DialectConversion.cpp`, so that it can be called from there in a future commit.

This commit is in preparation of decoupling argument/source/target materializations from the dialect conversion framework. The existing logic around unresolved materializations that predicts IR changes to decide if a cast op can be folded/erased will become obsolete, as `ReconcileUnrealizedCasts` will perform these kind of foldings on fully materialized IR.

### CODE REVIEW

**Comment by matthias-springer** [comment]
For reference, #104668 is still in development, but it's going to look somewhat like that.

File: mlir/include/mlir/Transforms/DialectConversion.h [hunk 1]
```diff
@@ -1126,6 +1126,29 @@ struct ConversionConfig {
...
+/// Example:
+/// %1 = unrealized_conversion_cast %0 : !A to !B
+/// %2 = unrealized_conversion_cast %1 : !B to !C
+/// %3 = unrealized_conversion_cast %2 : !C to !A
+///
+/// In the above example, %0 can be used instead of %3 and all cast ops are
+/// folded away.
+void reconcileUnrealizedCasts(
+    ArrayRef<UnrealizedConversionCastOp> castOps,
+    SmallVector<UnrealizedConversionCastOp> *remainingCastOps = nullptr);
```
Comment (zero9178) [inline mlir/include/mlir/Transforms/DialectConversion.h:1150]:
```suggestion
    SmallVectorImpl<UnrealizedConversionCastOp> *remainingCastOps = nullptr);
```

File: mlir/lib/Transforms/Utils/DialectConversion.cpp [hunk 2]
```diff
@@ -2869,6 +2869,80 @@ LogicalResult OperationConverter::legalizeErasedResult(
...
+        castOp->erase();
+        break;
+      }
+      nextCast = getInputCast(nextCast);
+    }
+  }
+
+  if (remainingCastOps)
+    for (UnrealizedConversionCastOp op : castOps)
+      if (!erasedOps.contains(op.getOperation()))
```
Comment (zero9178) [inline mlir/lib/Transforms/Utils/DialectConversion.cpp:2942]: This looks slightly scary given that `op` has technically been erased and `op.getOperation` now returns a dangling address. Can't come up with a better way of doing this either though and shouldn't be a correctness issue.

**Review by zero9178 — APPROVED** [review]

File: mlir/lib/Conversion/ReconcileUnrealizedCasts/ReconcileUnrealizedCasts.cpp [hunk 3]
```diff
@@ -39,63 +40,10 @@ struct ReconcileUnrealizedCasts
   ReconcileUnrealizedCasts() = default;
 
   void runOnOperation() override {
-    // Gather all unrealized_conversion_cast ops.
-    SetVector<UnrealizedConversionCastOp> worklist;
+    SmallVector<UnrealizedConversionCastOp> ops;
     getOperation()->walk(
         [&](UnrealizedConversionCastOp castOp) { worklist.insert(castOp); });
```
Comment (zero9178) [inline mlir/lib/Conversion/ReconcileUnrealizedCasts/ReconcileUnrealizedCasts.cpp:45]:
```suggestion
        [&](UnrealizedConversionCastOp castOp) { ops.insert(castOp); });
```

File: mlir/include/mlir/Transforms/DialectConversion.h (hunk 1 above)
Comment (matthias-springer) [reply mlir/include/mlir/Transforms/DialectConversion.h:1150]: What the difference between `SmallVector<...>` and `SmallVectorImpl<...>`?

**Pull request merged**